frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1124 @@
+	pub fn is_member(who: &T::AccountId) -> bool {
+		Self::member_role_of(who).is_some()
+	}
+
+	pub fn is_member_of(who: &T::AccountId, role: MemberRole) -> bool {
+		Members::<T, I>::get(role).contains(&who)
+	}
+
+	fn is_founder(who: &T::AccountId) -> bool {
+		Self::is_member_of(who, MemberRole::Founder)
+	}
+
+	fn is_fellow(who: &T::AccountId) -> bool {
+		Self::is_member_of(who, MemberRole::Fellow)
+	}
+
+	fn is_ally(who: &T::AccountId) -> bool {
+		Self::is_member_of(who, MemberRole::Ally)
+	}
+
+	fn is_votable_member(who: &T::AccountId) -> bool {
+		Self::is_founder(who) || Self::is_fellow(who)
+	}
+
+	fn votable_member_sorted() -> Vec<T::AccountId> {
+		let mut founders = Members::<T, I>::get(MemberRole::Founder).into_inner();
+		let mut fellows = Members::<T, I>::get(MemberRole::Fellow).into_inner();
+		founders.append(&mut fellows);
+		founders.sort();
+		founders.into()
+	}
+
+	fn is_kicking(who: &T::AccountId) -> bool {
+		<KickingMembers<T, I>>::contains_key(&who)
+	}
+
+	/// Add a user to the sorted alliance member set.
+	fn add_member(who: &T::AccountId, role: MemberRole) -> DispatchResult {
+		<Members<T, I>>::try_mutate(role, |members| -> DispatchResult {
+			let pos = members.binary_search(who).err().ok_or(Error::<T, I>::AlreadyMember)?;
+			members
+				.try_insert(pos, who.clone())
+				.map_err(|_| Error::<T, I>::TooManyMembers)?;
+			Ok(())
+		})?;
+
+		if role == MemberRole::Founder || role == MemberRole::Fellow {
+			let members = Self::votable_member_sorted();
+			T::MembershipChanged::change_members_sorted(&[who.clone()], &[], &members[..]);
+		}
+		Ok(())
+	}
+
+	/// Remove a user from the alliance member set.
+	fn remove_member(who: &T::AccountId, role: MemberRole) -> DispatchResult {
+		<Members<T, I>>::try_mutate(role, |members| -> DispatchResult {
+			let pos = members.binary_search(who).ok().ok_or(Error::<T, I>::NotMember)?;
+			members.remove(pos);
+			Ok(())
+		})?;
+
+		if role == MemberRole::Founder || role == MemberRole::Fellow {
+			let members = Self::votable_member_sorted();
+			T::MembershipChanged::change_members_sorted(&[], &[who.clone()], &members[..]);
+		}
+		Ok(())
+	}
+
+	/// Check if a user is in blacklist.
+	fn is_blacklist(info: &BlacklistItemOf<T, I>) -> bool {
+		match info {
+			BlacklistItem::Website(url) => <WebsiteBlacklist<T, I>>::get().contains(url),
+			BlacklistItem::AccountId(who) => <AccountBlacklist<T, I>>::get().contains(who),
+		}
+	}
+
+	/// Check if a user is in account blacklist.
+	fn is_account_blacklist(who: &T::AccountId) -> bool {
+		<AccountBlacklist<T, I>>::get().contains(who)
+	}
+
+	/// Add a identity info to the blacklist set.
+	fn do_add_blacklist(
+		new_accounts: &mut Vec<T::AccountId>,
+		new_webs: &mut Vec<UrlOf<T, I>>,
+	) -> DispatchResult {
+		if !new_accounts.is_empty() {
+			<AccountBlacklist<T, I>>::try_mutate(|accounts| -> DispatchResult {
+				accounts.try_append(new_accounts).map_err(|_| Error::<T, I>::TooManyBlacklist)?;
+				accounts.sort();
```

Comment:
Shouldn't front ends provide an index to place into? `sort` seems to be used heavily in this pallet.

Reply:
That could work but the implementation will be a bit complicated when adding multiple items at once.
Sort shouldn't be an issue if we don't have too many items and if we do, maybe we should use a map instead of sorted vec